### PR TITLE
Time Accounting (Part 6)

### PR DESCRIPTION
**Pull request by swalker2m**
Current state of time accounting, now including:

* Storage of time accounting data
* Discount for daylight (pre-dusk and/or post-dawn)
* Discount when no data is collected
* Discount for QA state
* TimeAccountingState updates
* Observation and Program time charge sum
* Scanty testing

Incorporates #908 #905 #900 #896 #885.

### CODE REVIEW

File: modules/service/src/main/resources/db/migration/V0839__time_discount.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+-- Time accounting charge discounts
+CREATE TABLE t_time_charge_discount (
+
+  c_id               bigserial   PRIMARY KEY,
+
+  c_visit_id         d_visit_id  NOT NULL REFERENCES t_visit(c_visit_id) ON DELETE CASCADE,
+
+  c_start            timestamp   NOT NULL,
+
+  c_end              timestamp   NOT NULL,
```
Comment (tpolecat) [inline modules/service/src/main/resources/db/migration/V0839__time_discount.sql:19]: Do these need to be constrained such that `c_start` <= `c_end`?

File: modules/service/src/main/scala/lucuma/odb/service/TimeAccountingService.scala [hunk 2]
```diff
@@ -3,59 +3,341 @@
...
+              // ... other discounts here ...
+              fin <- StateT.get[F, TimeAccountingState]
+            } yield TimeCharge.Invoice(ini.charge, nod ++ day ++ qa, fin.charge.corrected(corrections))
 
+          for {
+            s <- initialState
+            c <- session.execute(SelectCorrection)(visitId)
+            i <- invoice(c).runA(s)
+          } yield i
+        }
```
Comment (tpolecat) [inline modules/service/src/main/scala/lucuma/odb/service/TimeAccountingService.scala:210]: Nicely done here.

**Review by tpolecat — APPROVED** [review]
This is a big PR but the code is very clear. 👍🏻

File: modules/service/src/main/resources/db/migration/V0839__time_discount.sql (hunk 1 above)
Comment (swalker2m) [reply modules/service/src/main/resources/db/migration/V0839__time_discount.sql:19]: Yes, thanks, I overlooked that one.

**Pull request merged**